### PR TITLE
Update design for exportable subflow

### PR DESCRIPTION
**Pull request by HiroyasuNishiyama**
This PR is for updating design note for exportable subflow (called SUBFLOW Module) based on discussions held on early July.

It includes following items:
- update for export interface,
- deleting imported subflow module,
- representation of subflow module.

### CODE REVIEW

File: designs/exportable-subflow/README.md [hunk 1]
```diff
@@ -24,28 +24,52 @@ are in the roadmap. (*TODO: add a link to the design doc once added*).
...
 
 ![Subflow-export-UI](Subflow-export-UI.png)
 
-By clicking **Export** button on Subflow Setting Panel of Subflow template, Subflow Export Panel is shown.
-
-In this panel, new node information such as node name, version, etc. can be specified. There is **As JSON** check box on this panel.  If this is checked, JSON export panel is shown.  Otherwise, new node is installed on  editor palette.  
+By clicking **Export SUBFLOW** button on SUBFLOW template setting panel, SUBFLOW export panel is shown.
 
-New node created from SUBFLOW is sealed.  Thus, its internal flow or UI definition can not be edited.
+In this panel, new node information such as node name, version, etc. can be specified. There is **Include All SUBFLOW** check box on this panel.  If this is checked, SUBFLOWs used within the exported SUBFLOW are recursively included in exported SUBFLOW Module.  If **Editable** check box on this panel is checked, SUBFLOW template can be edited on imported Node-RED editor.  Otherwise, the imported SUBLOW template is sealed. Thus, its internal flow or UI definition can not be edited.
```
Comment (knolleary) [inline designs/exportable-subflow/README.md]: I think *most* of this extra information should be on a new tab in the subflow *edit* dialog that is saved as properties of the subflow regardless of the Export work flow.

If I am the author of a subflow that I want to distribute, I will have a local Node-RED project where I develop and test the subflow. I then export it and share with others.
If I then need to make an update to the subflow (such as a bug fix), I do not want to have to re-enter all of these extra properties when I re-export. That could easily lead to a mistake being made. I should be able to reopen my Node-RED project, make the change to the subflow and re-export.

File: designs/exportable-subflow/README.md [hunk 2]
```diff
@@ -24,28 +24,52 @@ are in the roadmap. (*TODO: add a link to the design doc once added*).
...
 
-Node represented as JSON data can be imported in Node-RED environment from palette management UI.
+Node represented as JSON data can be imported in Node-RED environment from import new node panel.
 
 ![Subflow-import-UI](Subflow-import-UI.png)
 
-There is **Install(JSON)** button on top of the panel.  After importing JSON, new node is installed on editor palette.
+#### Uninstalling SUBFLOW Module
+
+SUBFLOW Module are listed in *Manage Palette > Palette > Nodes* tab.  Similar to normal nodes, *SUBFLOW Module*s listed in this tab will have **remove** button.  If a SUBFLOW Module includes other SUBFLOW Modules and is sealed, contained SUBFLOW Modules are also deleted.
```
Comment (knolleary) [inline designs/exportable-subflow/README.md]: Does this apply to all subflows? Or only sealed ones? Non-sealed subflows can be deleted in the usual way.

File: designs/exportable-subflow/README.md [hunk 3]
```diff
@@ -24,28 +24,52 @@ are in the roadmap. (*TODO: add a link to the design doc once added*).
...
+
+Exported SUBFLOW Module is represented by JSON array format which is similar to current flow representation.  
+
+Following is a list of properties exported SUBFLOW node in SUBFLOW Module may contain (+ means that property is optional):
+
+|      | name    | type   | description                        |
+| ---- | ------- | ------ | ---------------------------------- |
+| 1    | flow    | array  | array of contained nodes in SUBFLOW |
+| 2    | sealed+ | bool   | SUBFLOW template is not editable   |
+| 3    | hidden+ | bool   | SUBFLOW template is hidden         |
```
Comment (knolleary) [inline designs/exportable-subflow/README.md]: What is the `hidden` property for? I can't remember discussing that (apologies if we did) but this design doesn't explain what it means.

File: designs/exportable-subflow/README.md [hunk 4]
```diff
@@ -24,28 +24,52 @@ are in the roadmap. (*TODO: add a link to the design doc once added*).
...
+| 5    | color+  | string | color of SUBFLOW                   |
+| 6    | version | string | version of SUBFLOW (format: x.y.z) |
+| 7    | author+ | string | author of SUBFLOW                  |
+| 8    | license+ | string | license of SUBFLOW                 |
+| 9    | keywords+ | array  | array of keywords of SUBFLOW       |
+| 10   | description+ | string | simple description of SUBFLOW      |
+
+Exported SUBFLOW Module can be **sealed** which means that internal of the SUBFLOW Module is hidden from users.  
+
+`flow` property is an array of nodes that are used in its SUBFLOW template.  *id* of nodes in SUBFLOW module must be *name*-*ID* where *ID* is a original *id* of the node.
```
Comment (knolleary) [inline designs/exportable-subflow/README.md]: Where has this requirement for the ids of the nodes inside the subflow come from? What does it solve doing this?

**Comment by HiroyasuNishiyama** [comment]
Updated document according to the last teleconference.

File: designs/exportable-subflow/README.md (hunk 1 above)
Comment (HiroyasuNishiyama) [reply designs/exportable-subflow/README.md]: Moved properties edit entries to SUBFLOW Template properties edit tab.

File: designs/exportable-subflow/README.md (hunk 2 above)
Comment (HiroyasuNishiyama) [reply designs/exportable-subflow/README.md]: Current proposal apply to imported one.

File: designs/exportable-subflow/README.md (hunk 3 above)
Comment (HiroyasuNishiyama) [reply designs/exportable-subflow/README.md]: Add description of `hidden` property to design note.
It is used for SUBFLOW contained in sealed SUBFLOW.

File: designs/exportable-subflow/README.md (hunk 4 above)
Comment (HiroyasuNishiyama) [reply designs/exportable-subflow/README.md]: Not discussed this issue on teleconference.
This is for different flows that use same SUBFLOW Template.
If ID is random string, existing flow can not refer to imported SUBFLOW Template.

**Comment by knolleary** [comment]
Looks good for the first couple of phases.

I suspect will need some more design discussion around sealed subflows - there's a lot more to consider there around what differences in behaviour are expected for a sealed subflow. Let's do that as a new PR on this design as we get closer to that point.

**Pull request merged**